test/python/circuit/test_control_flow_builders.py

```diff
@@ -0,0 +1,2191 @@
+    def test_break_continue_accept_c_if(self, loop_operation):
+        """Test that ``break`` and ``continue`` statements accept :meth:`.Instruction.c_if` calls,
+        and that these propagate through correctly."""
+        qubits = [Qubit(), Qubit()]
+        clbits = [Clbit(), Clbit()]
+
+        with self.subTest("for"):
+            test = QuantumCircuit(qubits, clbits)
+            with test.for_loop(None, range(2)):
+                test.h(0)
+                loop_operation(test).c_if(1, 0)
+
+            body = QuantumCircuit([qubits[0]], [clbits[1]])
+            body.h(qubits[0])
+            loop_operation(body).c_if(clbits[1], 0)
+
+            expected = QuantumCircuit(qubits, clbits)
+            expected.for_loop(None, range(2), body, [qubits[0]], [clbits[1]])
+
+            self.assertCircuitsEquivalent(test, expected)
+
+        with self.subTest("while"):
+            cond = (clbits[0], 0)
+            test = QuantumCircuit(qubits, clbits)
+            with test.while_loop(cond):
+                test.h(0)
+                loop_operation(test).c_if(1, 0)
+
+            body = QuantumCircuit([qubits[0]], clbits)
+            body.h(qubits[0])
+            loop_operation(body).c_if(clbits[1], 0)
+
+            expected = QuantumCircuit(qubits, clbits)
+            expected.while_loop(cond, body, [qubits[0]], clbits)
+
+            self.assertCircuitsEquivalent(test, expected)
+
+    @ddt.data(QuantumCircuit.break_loop, QuantumCircuit.continue_loop)
+    def test_break_continue_only_expand_to_nearest_loop(self, loop_operation):
+        """Test that a ``break`` or ``continue`` nested in more than loop only expands as far as the
+        inner loop scope, not further."""
+        qubits = [Qubit(), Qubit(), Qubit()]
+        clbits = [Clbit(), Clbit(), Clbit()]
+        cond_inner = (clbits[0], 0)
+        cond_outer = (clbits[1], 0)
+
+        with self.subTest("for for"):
+            test = QuantumCircuit(qubits, clbits)
+            with test.for_loop(None, range(2)):
+                test.measure(1, 1)
+                with test.for_loop(None, range(2)):
+                    test.h(0)
+                    loop_operation(test)
+                loop_operation(test)
+
+            inner_body = QuantumCircuit([qubits[0]])
+            inner_body.h(qubits[0])
+            loop_operation(inner_body)
+
+            outer_body = QuantumCircuit([qubits[0], qubits[1], clbits[1]])
+            outer_body.measure(qubits[1], clbits[1])
+            outer_body.for_loop(None, range(2), inner_body, [qubits[0]], [])
+            loop_operation(outer_body)
+
+            expected = QuantumCircuit(qubits, clbits)
+            expected.for_loop(None, range(2), outer_body, [qubits[0], qubits[1]], [clbits[1]])
+
+            self.assertCircuitsEquivalent(test, expected)
+
+        with self.subTest("for while"):
+            test = QuantumCircuit(qubits, clbits)
+            with test.for_loop(None, range(2)):
```

Comment:
Wow! It's immediately obvious just how much value the control builder interface adds to the usability of programming control flow logic. It's _very_ difficult to at a glance understand the `expected` circuits in this file, comparatively.

Reply:
Yeah, it was a straight-up nightmare writing them :(.  I had _sooo_ many failing tests repeatedly, and I think it was about twice that the builders were wrong, and fifty that I'd written the expected test wrong.